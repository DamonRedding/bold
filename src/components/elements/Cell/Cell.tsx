import { Interpolation } from 'emotion'
import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../../styles'
import { Omit } from '../../../util'

export type AlignSelf = 'auto' | 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch'
export type CellSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12

export interface CellProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  size?: CellSize
  style?: Interpolation
  alignSelf?: AlignSelf
  flexGrow?: CSSProperties['flexGrow']
  flexShrink?: CSSProperties['flexShrink']
  flexBasis?: CSSProperties['flexBasis']
}

export function Cell(props: CellProps) {
  const { size, style, alignSelf, flexGrow, flexShrink, flexBasis, ...rest } = props
  const { classes, css } = useStyles(createStyles, props)

  const className = css(classes.cell, size && classes.fixedSize, style)

  return <div className={className} {...rest} />
}

export const createStyles = (theme: Theme, props: CellProps) => ({
  cell: {
    margin: '0.5rem 1rem',
    alignSelf: props.alignSelf,
    flexBasis: props.flexBasis,
    flexGrow: props.flexGrow,
    flexShrink: props.flexShrink,

    [theme.breakpoints.down('small')]: {
      margin: '0.25rem 0.5rem',
    },
  } as React.CSSProperties,

  fixedSize: {
    width: `calc((100% / 12 * ${props.size}) - 2rem)`,

    [theme.breakpoints.down('small')]: {
      width: `calc((100% / 12 * ${props.size}) - 1rem)`,
    },
  },
})
