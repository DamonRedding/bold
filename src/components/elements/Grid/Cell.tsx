import { Interpolation } from 'emotion'
import React, { CSSProperties, useContext } from 'react'

import { Theme, useStyles } from '../../../styles'
import { Omit } from '../../../util'

import { Grid, GridContext, GridProps } from './Grid'

export type AlignSelf = 'auto' | 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch'
export type CellSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12

export interface CellProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  style?: Interpolation
  alignSelf?: AlignSelf
  flexGrow?: CSSProperties['flexGrow']
  flexShrink?: CSSProperties['flexShrink']
  flexBasis?: CSSProperties['flexBasis']

  /**
   * Width of the cell, defined using a 12-column based grid (12 is 100%, 6 is 50%, 4 is 25%, etc.).
   */
  size?: CellSize

  /**
   * Width to be used on extra-small (xs) or larger devices, based on breakpoints defined on theme.
   * The size is defined using a 12-column based grid (12 is 100%, 6 is 50%, 4 is 25%, etc.).
   */
  xs?: CellSize

  /**
   * Width to be used on small (sm) or larger devices, based on breakpoints defined on theme.
   * The size is defined using a 12-column based grid (12 is 100%, 6 is 50%, 4 is 25%, etc.).
   */
  sm?: CellSize

  /**
   * Width to be used on medium (md) or larger devices, based on breakpoints defined on theme.
   * The size is defined using a 12-column based grid (12 is 100%, 6 is 50%, 4 is 25%, etc.).
   */
  md?: CellSize

  /**
   * Width to be used on large (lg) or larger devices, based on breakpoints defined on theme.
   * The size is defined using a 12-column based grid (12 is 100%, 6 is 50%, 4 is 25%, etc.).
   */
  lg?: CellSize

  /**
   * Width to be used on extra-large (xl) or larger devices, based on breakpoints defined on theme.
   * The size is defined using a 12-column based grid (12 is 100%, 6 is 50%, 4 is 25%, etc.).
   */
  xl?: CellSize
}

export function Cell(props: CellProps) {
  const { size, style, alignSelf, flexGrow, flexShrink, flexBasis, xs, sm, md, lg, xl, ...rest } = props
  const gridProps = useContext(GridContext) || Grid.defaultProps
  const { classes, css } = useStyles(createStyles, props, gridProps)

  const className = css(
    classes.cell,
    (xs || size) && classes.xs,
    sm && classes.sm,
    md && classes.md,
    lg && classes.lg,
    xl && classes.xl,
    style
  )

  return <div className={className} {...rest} />
}

export const createStyles = (
  theme: Theme,
  { alignSelf, flexBasis, flexGrow, flexShrink, size, xs, sm, md, lg, xl }: CellProps,
  { gap, gapCrossAxis, direction }: GridProps
) => ({
  cell: {
    margin: ['column', 'column-reverse'].includes(direction)
      ? `${gap / 2}rem ${gapCrossAxis / 2}rem`
      : `${gapCrossAxis / 2}rem ${gap / 2}rem`,
    alignSelf,
    flexBasis,
    flexGrow,
    flexShrink,
  } as React.CSSProperties,

  xs: {
    width: `calc((100% / 12 * ${xs || size}) - ${gap}rem)`,
  },
  sm: {
    [theme.breakpoints.up('sm')]: {
      width: `calc((100% / 12 * ${sm}) - ${gap}rem)`,
    },
  },
  md: {
    [theme.breakpoints.up('md')]: {
      width: `calc((100% / 12 * ${md}) - ${gap}rem)`,
    },
  },
  lg: {
    [theme.breakpoints.up('lg')]: {
      width: `calc((100% / 12 * ${lg}) - ${gap}rem)`,
    },
  },
  xl: {
    [theme.breakpoints.up('xl')]: {
      width: `calc((100% / 12 * ${xl}) - ${gap}rem)`,
    },
  },
})
