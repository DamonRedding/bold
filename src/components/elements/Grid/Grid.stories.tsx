import { number, select } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import { Interpolation } from 'emotion'
import React from 'react'

import { AlignItems, AlignSelf, Cell, Direction, Grid, JustifyContent } from '../'

const styles = {
  box: {
    backgroundColor: '#eee',
    textAlign: 'center',
  } as Interpolation,
}

const justifyContentOptions: JustifyContent[] = [
  'flex-start',
  'flex-end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
]
const alignItemsOptions: AlignItems[] = ['flex-start', 'flex-end', 'center', 'baseline', 'stretch']
const alignSelfOptions: AlignSelf[] = ['auto', 'flex-start', 'flex-end', 'center', 'baseline', 'stretch']
const directionOptions: Direction[] = ['row', 'row-reverse', 'column', 'column-reverse']

storiesOf('Components|Grid', module)
  .add('default', () => (
    <Grid>
      <Cell size={6} style={styles.box}>
        size=6
      </Cell>
      <Cell size={3} style={styles.box}>
        size=3
      </Cell>
      <Cell size={2} style={styles.box}>
        size=2
      </Cell>
      <Cell size={1} style={styles.box}>
        size=1
      </Cell>
    </Grid>
  ))
  .add('flex-grow', () => (
    <Grid>
      <Cell flexGrow={1} style={styles.box}>
        1
      </Cell>
      <Cell style={styles.box}>2</Cell>
      <Cell flexGrow={2} style={styles.box}>
        3
      </Cell>
      <Cell style={styles.box}>4</Cell>
    </Grid>
  ))
  .add('playground', () => (
    <>
      <Grid
        alignItems={select('alignItems', alignItemsOptions, 'center')}
        justifyContent={select('justifyContent', justifyContentOptions, 'center')}
        direction={select('direction', directionOptions, 'row')}
        gap={number('gap', 2)}
        gapCrossAxis={number('gapCrossAxis', 1)}
        wrap
      >
        <Cell size={6} style={styles.box}>
          size=6
        </Cell>
        <Cell size={6} style={styles.box}>
          size=6
        </Cell>
        <Cell size={6} style={styles.box}>
          size=6
        </Cell>
      </Grid>
      <hr />
      <div style={{ height: 300 }}>
        <Grid
          alignItems={select('alignItems', alignItemsOptions, 'center')}
          justifyContent={select('justifyContent', justifyContentOptions, 'center')}
          direction={select('direction', directionOptions, 'row')}
          style={{ height: '100%' }}
        >
          <Cell size={1} style={styles.box}>
            1
          </Cell>
          <Cell size={1} style={styles.box}>
            <div style={{ height: 60 }}>2</div>
          </Cell>
          <Cell size={1} style={styles.box}>
            3
          </Cell>
        </Grid>
      </div>
    </>
  ))
  .add('cell', () => (
    <Grid alignItems='flex-start'>
      <Cell size={1} style={styles.box}>
        1
      </Cell>
      <Cell size={1} style={styles.box}>
        <div style={{ height: 60 }}>2</div>
      </Cell>
      <Cell
        size={number('size', 2) as any}
        alignSelf={select('alignSelf', alignSelfOptions, 'auto')}
        style={styles.box}
      >
        <strong>3</strong>
      </Cell>
      <Cell size={1} style={styles.box}>
        4
      </Cell>
    </Grid>
  ))
