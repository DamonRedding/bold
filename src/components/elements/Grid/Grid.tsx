import { Interpolation } from 'emotion'
import React from 'react'

import { Theme, useStyles } from '../../../styles'
import { Omit } from '../../../util'

export type AlignItems = 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch'
export type JustifyContent = 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around' | 'space-evenly'
export type Direction = 'row' | 'row-reverse' | 'column' | 'column-reverse'

export interface GridProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  wrap?: boolean
  alignItems?: AlignItems
  justifyContent?: JustifyContent
  direction?: Direction
  style?: Interpolation
}

export const Grid = (props: GridProps) => {
  const { style, wrap, alignItems, justifyContent, direction, ...rest } = props
  const { classes, css } = useStyles(createStyles, props)

  const className = css(classes.grid, wrap && classes.wrap, style)

  return <div className={className} {...rest} />
}

export const createStyles = (theme: Theme, props: GridProps) => ({
  grid: {
    alignItems: props.alignItems,
    display: 'flex',
    flexDirection: props.direction,
    justifyContent: props.justifyContent,
    marginLeft: '-1rem',
    marginRight: '-1rem',
    height: '100%',

    [theme.breakpoints.down('small')]: {
      marginLeft: '-0.5rem',
      marginRight: '-0.5rem',
    },
  } as React.CSSProperties,
  wrap: {
    flexWrap: 'wrap',
  } as React.CSSProperties,
})
