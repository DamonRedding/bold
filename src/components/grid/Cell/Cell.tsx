import * as React from 'react'

import { makeOptionClasses, Theme, withStyles, WithStylesProps } from '../../../styles'

export type AlignSelf = 'auto' | 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch'

export interface CellProps extends WithStylesProps {
    size?: number
    alignSelf?: AlignSelf
}

const createCellStyle = (theme: Theme, size: number) => {
    return {
        flexBasis: `calc((100% / 12 * ${size}) - 2rem)`,
        [theme.breakpoint.small]: {
            flexBasis: `calc((100% / 12 * ${size}) - 1rem)`,
        },
    }
}

@withStyles
export class Cell extends React.PureComponent<CellProps> {
    render() {
        const styles = this.props.createStyles(theme => ({
            cell: {
                margin: '1rem',
                padding: '0.5rem',

                [theme.breakpoint.small]: {
                    margin: '0.5rem',
                },
            },

            autoSize: {
                flexGrow: 1,
            },

            size1: createCellStyle(theme, 1),
            size2: createCellStyle(theme, 2),
            size3: createCellStyle(theme, 3),
            size4: createCellStyle(theme, 4),
            size5: createCellStyle(theme, 5),
            size6: createCellStyle(theme, 6),
            size7: createCellStyle(theme, 7),
            size8: createCellStyle(theme, 8),
            size9: createCellStyle(theme, 9),
            size10: createCellStyle(theme, 10),
            size11: createCellStyle(theme, 11),
            size12: createCellStyle(theme, 12),

            ...makeOptionClasses('alignSelf__', 'alignSelf',
                ['auto', 'flex-start', 'flex-end', 'center', 'baseline', 'stretch']),
        }))

        const classes = this.props.css(
            styles.cell,
            this.props.size ? styles[`size${this.props.size}`] : styles.autoSize,
            this.props.alignSelf && styles['alignSelf__' + this.props.alignSelf]
        )

        return (
            <div className={classes}>
                {this.props.children}
            </div>
        )
    }
}
