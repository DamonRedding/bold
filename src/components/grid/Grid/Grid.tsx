import * as React from 'react'

import { withStyles, WithStylesProps } from '../../../styles'

export type AlignItems = 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch'
export type JustifyContent = 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around' | 'space-evenly'
export type Direction = 'row' | 'row-reverse' | 'column' | 'column-reverse'

export interface GridProps extends WithStylesProps, Pick<React.CSSProperties, 'alignItems' | 'justifyContent'> {
    wrap?: boolean
    alignItems?: AlignItems
    justifyContent?: JustifyContent
    direction?: Direction
}

@withStyles
export class Grid extends React.PureComponent<GridProps> {
    render() {
        const { theme } = this.props
        const styles = {
            grid: {
                alignItems: this.props.alignItems,
                display: 'flex',
                flexDirection: this.props.direction,
                justifyContent: this.props.justifyContent,
                marginLeft: '-1rem',
                marginRight: '-1rem',
                height: '100%',

                [theme.breakpoints.down('small')]: {
                    marginLeft: '-0.5rem',
                    marginRight: '-0.5rem',
                },
            },

            wrap: {
                flexWrap: 'wrap',
            },
        }

        const classes = this.props.css(
            styles.grid,
            this.props.wrap && styles.wrap
        )

        return (
            <div className={classes}>
                {this.props.children}
            </div>
        )
    }
}
