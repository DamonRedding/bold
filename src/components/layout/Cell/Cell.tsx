import { Interpolation } from 'emotion'
import React from 'react'

import { withStyles, WithStylesProps } from '../../../styles'

export type AlignSelf = 'auto' | 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch'
export type CellSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12

export interface CellProps extends WithStylesProps {
    size?: CellSize
    alignSelf?: AlignSelf
    style?: Interpolation
}

@withStyles
export class Cell extends React.PureComponent<CellProps> {
    render() {
        const { size, theme, style } = this.props
        const styles = {
            cell: {
                alignSelf: this.props.alignSelf,
                margin: '0.5rem 1rem',

                [theme.breakpoints.down('small')]: {
                    margin: '0.25rem 0.5rem',
                },
            },

            autoSize: {
                flexGrow: 1,
            },

            fixedSize: {
                flexBasis: `calc((100% / 12 * ${size}) - 2rem)`,
                [theme.breakpoints.down('small')]: {
                    flexBasis: `calc((100% / 12 * ${size}) - 1rem)`,
                },
            },

        }

        const classes = this.props.css(
            styles.cell,
            this.props.size ? styles.fixedSize : styles.autoSize,
            style
        )

        return (
            <div className={classes}>
                {this.props.children}
            </div>
        )
    }
}
