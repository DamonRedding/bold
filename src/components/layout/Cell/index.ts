export { Cell, CellProps, AlignSelf, CellSize } from './Cell'
