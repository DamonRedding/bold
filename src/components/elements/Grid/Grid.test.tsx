import React from 'react'
import { render } from 'react-testing-library'

import { Cell } from './Cell'
import { Grid, GridProps } from './Grid'

it('should render correctly', () => {
  expect(render(<Grid>Content</Grid>).container).toMatchSnapshot()
})

it('should accept flex container props', () => {
  const { container } = render(
    <Grid justifyContent='space-around' alignItems='center' direction='column-reverse' wrap>
      Content
    </Grid>
  )
  expect(container).toMatchSnapshot()
})

it('should accept "style" prop', () => {
  const { container } = render(<Grid style={{ color: 'red' }}>Content</Grid>)
  expect(container).toMatchSnapshot()
})

it('should accept "gap" and "gapVertical" props', () => {
  const createGrid = (props: Partial<GridProps> = {}) => (
    <Grid gap={10} gapVertical={5} {...props}>
      <Cell>1</Cell>
      <Cell>2</Cell>
    </Grid>
  )

  const { container, rerender } = render(createGrid())
  expect(container).toMatchSnapshot()

  rerender(createGrid({ direction: 'row-reverse' }))
  expect(container).toMatchSnapshot()

  rerender(createGrid({ direction: 'column' }))
  expect(container).toMatchSnapshot()

  rerender(createGrid({ direction: 'column-reverse' }))
  expect(container).toMatchSnapshot()
})

it('should pass props down to div', () => {
  const { container } = render(
    <Grid id='div' aria-label='test'>
      Content
    </Grid>
  )
  expect(container.querySelector('div').getAttribute('id')).toEqual('div')
  expect(container.querySelector('div').getAttribute('aria-label')).toEqual('test')
})
