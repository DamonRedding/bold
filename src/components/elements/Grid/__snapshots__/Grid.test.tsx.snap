// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "style" prop 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -1rem;
  margin-right: -1rem;
  height: 100%;
  color: red;
}

@media (max-width:1280px) {
  .emotion-0 {
    margin-left: -0.5rem;
    margin-right: -0.5rem;
  }
}

<div>
  <div
    class="emotion-0"
  >
    Content
  </div>
</div>
`;

exports[`should accept flex container props 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  margin-left: -1rem;
  margin-right: -1rem;
  height: 100%;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

@media (max-width:1280px) {
  .emotion-0 {
    margin-left: -0.5rem;
    margin-right: -0.5rem;
  }
}

<div>
  <div
    class="emotion-0"
  >
    Content
  </div>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: -1rem;
  margin-right: -1rem;
  height: 100%;
}

@media (max-width:1280px) {
  .emotion-0 {
    margin-left: -0.5rem;
    margin-right: -0.5rem;
  }
}

<div>
  <div
    class="emotion-0"
  >
    Content
  </div>
</div>
`;
